### ISSUE TITLE
Enhancement: [no-base-to-string] Add option to flag functions

### ISSUE THREAD

**Issue by SuttonKyle** [typescript-eslint/typescript-eslint#7337]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-base-to-string/

### Description

I propose that an option should be added to `no-base-to-string` to cause it to flag functions. In most use cases, stringifying a function is not desirable and usually unexpected behavior, so the current carveout for functions and other values that define a `toString()` method doesn't make a ton of sense for a lot of projects. Adding a boolean option like `rejectFunctions` or `flagFunctions` would make it possible to flag these unexpected behaviors for developers that want the option.

### Fail

```typescript
const fn = () => "hello world!";

`${fn}`;
fn.toString();
```

### Pass

```typescript
const fn = () => "hello world!";

`${fn()}`;
```

### Additional Info

I've implemented this option already in a forked repository, so I'm ready to PR if the proposal gets accepted.

**Comment by bradzacher** [comment]
I don't think this belongs within `no-base-to-string`. As defined - the function prototype includes a `toString` definition - thus it should pass the rule.

I could see this as a new rule as it seems like a pretty valuable standalone check, rather than being obfuscated behind an out-of-place option.

**Comment by SuttonKyle** [comment]
Resolved -- the functionality we need exists in `restrict-template-expressions` and `restrict-plus-operands`
